### ISSUE TITLE
Support really large models when a submodule size is close to GPU memory

### ISSUE THREAD

**Issue by pbelevich** [pytorch/tau#298]
Currently we create a model either on GPU or CPU memory, trace it and split it into submodules and all of this fits into memory. For really large models(roughly >10B params) that cannot fit into memory we should create a model on meta device, trace it and split it on meta device. Then we have two options: 
1. Materialize each submodule on the master node and send over RPC to a worker. Because of potential lack of memory this should be done sequentially for each submodule. This doesn't require any PyTorch RPC changes since the submodule is materialized either on CPU or GPU. Although it may require per-RPC device mapping.
2. Implement meta device support in PyTorch RPC and send meta-device submodules to the workers. This can be done in parallel. Each worker will materialize its submodule on its own.

Option 1 can be implemented first and then we can work on option 2 which is preferable.

Open question is how to properly initialize submodules during materialization.

**Comment by pbelevich** [comment]
blocked by PyTorch RPC meta/fake tensors support

**Comment by pbelevich** [comment]
RPC Meta device support: https://github.com/pytorch/pytorch/pull/82530

**Comment by pbelevich** [comment]
PyTorch core FakeTensor RPC error:
```
Traceback (most recent call last):
  File "/home/pbelevich/local/anaconda3/envs/pt-dev/lib/python3.9/site-packages/torch/multiprocessing/spawn.py", line 69, in _wrap
    fn(i, *args)
  File "/home/pbelevich/local/PiPPy/test/test_rcp_fake.py", line 55, in run_worker
    run_master(args)
  File "/home/pbelevich/local/PiPPy/test/test_rcp_fake.py", line 26, in run_master
    rpc.rpc_sync("worker1", noop, args=(torch.empty(42, device='cpu'),))
  File "/home/pbelevich/local/anaconda3/envs/pt-dev/lib/python3.9/site-packages/torch/distributed/rpc/api.py", line 83, in wrapper
    return func(*args, **kwargs)
  File "/home/pbelevich/local/anaconda3/envs/pt-dev/lib/python3.9/site-packages/torch/distributed/rpc/api.py", line 799, in rpc_sync
    fut = _invoke_rpc(to, func, RPCExecMode.SYNC, args, kwargs, timeout)
  File "/home/pbelevich/local/anaconda3/envs/pt-dev/lib/python3.9/site-packages/torch/distributed/rpc/api.py", line 708, in _invoke_rpc
    (pickled_python_udf, tensors) = _default_pickler.serialize(
  File "/home/pbelevich/local/anaconda3/envs/pt-dev/lib/python3.9/site-packages/torch/distributed/rpc/internal.py", line 133, in serialize
    p.dump(obj)
AttributeError: Can't pickle local object 'WeakValueDictionary.__init__.<locals>.remove'
```

**Comment by pbelevich** [comment]
torchdistx FakeTensor RPC error:
```
terminate called after throwing an instance of 'c10::NotImplementedError'
  what():  Cannot access storage of FakeTensorImpl
Exception raised from throw_storage_access_error at ../c10/core/TensorImpl.cpp:360 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x3e (0x7fef95b3e50e in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #1: c10::TensorImpl::throw_storage_access_error() const + 0x8b (0x7fef95b172fd in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #2: <unknown function> + 0x17959 (0x7feeed474959 in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torchdistx/lib/libtorchdistx.so)
frame #3: torch::jit::Pickler::pushStorageOfTensor(at::Tensor const&) + 0x152 (0x7fefc3c80ce2 in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #4: torch::jit::Pickler::pushLiteralTensor(c10::IValue const&) + 0x13d (0x7fefc3c8146d in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #5: torch::jit::Pickler::pushIValueImpl(c10::IValue const&) + 0x95 (0x7fefc3c81c15 in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #6: torch::jit::Pickler::pushIValue(c10::IValue const&) + 0x1b3 (0x7fefc3c82943 in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #7: <unknown function> + 0x4510cfe (0x7fefc3c82cfe in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #8: torch::jit::Pickler::pushSpecializedList(c10::IValue const&, char const*, std::function<void (c10::IValue const&)> const&) + 0x103 (0x7fefc3c80973 in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #9: torch::jit::Pickler::pushIValueImpl(c10::IValue const&) + 0x170 (0x7fefc3c81cf0 in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #10: torch::jit::Pickler::pushIValue(c10::IValue const&) + 0x63 (0x7fefc3c827f3 in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #11: torch::distributed::rpc::tensorpipeSerialize(c10::intrusive_ptr<torch::distributed::rpc::Message, c10::detail::intrusive_target_default_null_type<torch::distributed::rpc::Message> >, std::vector<c10::Device, std::allocator<c10::Device> >, std::vector<c10::Stream, std::allocator<c10::Stream> > const&) + 0xc2f (0x7fefc41aeb1f in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #12: torch::distributed::rpc::TensorPipeAgent::pipeWrite(std::shared_ptr<tensorpipe::Pipe> const&, c10::intrusive_ptr<torch::distributed::rpc::Message, c10::detail::intrusive_target_default_null_type<torch::distributed::rpc::Message> >, std::vector<c10::Device, std::allocator<c10::Device> >&&, std::vector<c10::Stream, std::allocator<c10::Stream> >, std::function<void (tensorpipe::Error const&)>) + 0xf6 (0x7fefc419c446 in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #13: torch::distributed::rpc::TensorPipeAgent::send(torch::distributed::rpc::WorkerInfo const&, c10::intrusive_ptr<torch::distributed::rpc::Message, c10::detail::intrusive_target_default_null_type<torch::distributed::rpc::Message> >, float, std::unordered_map<c10::Device, c10::Device, std::hash<c10::Device>, std::equal_to<c10::Device>, std::allocator<std::pair<c10::Device const, c10::Device> > > const&) + 0xe33 (0x7fefc41a3423 in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #14: torch::distributed::autograd::sendMessageWithAutograd(torch::distributed::rpc::RpcAgent&, torch::distributed::rpc::WorkerInfo const&, c10::intrusive_ptr<torch::distributed::rpc::Message, c10::detail::intrusive_target_default_null_type<torch::distributed::rpc::Message> >, bool, float, bool) + 0x12c (0x7fefc413756c in /home/pbelevich/local/anaconda3/envs/pippy/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)

```

**Comment by kwen2501** [comment]
Supported by https://github.com/pytorch/tau/pull/627.
Instead of adding RPC support for Fake tensors, we allow each rank to trace the full model so that it can retrieve and materialize only its own stage.
Closing.
